### DISCUSSION TITLE
Get and Set values for object form

### DISCUSSION THREAD

**Discussion by WeylonSantana** [rsuite/rsuite#3100] [Q&A]
Hi everyone, I'm new to rsuite, so I'm having problems with the form.
I have a property is that of type object
Consider something like

```typescript
//non-nested-properties above...
Security: ObjectType().shape({
  IpBlacklist: ArrayType().of(StringType().isRequired(RequiredString)),
  Packets: ObjectType().shape({
    EditorFloodThreshholds: ObjectType().shape({
      MaxPacketSize: NumberType().min(0).isRequired(RequiredString),
      MaxPacketPerSec: NumberType().min(0).isRequired(RequiredString),
      KickAvgPacketPerSec: NumberType().min(0).isRequired(RequiredString),
    }),
    PlayerFloodThreshholds: ObjectType().shape({
      MaxPacketSize: NumberType().min(0).isRequired(RequiredString),
      MaxPacketPerSec: NumberType().min(0).isRequired(RequiredString),
      KickAvgPacketPerSec: NumberType().min(0).isRequired(RequiredString),
    }),
    ModAdminFloodThreshholds: ObjectType().shape({
      MaxPacketSize: NumberType().min(0).isRequired(RequiredString),
      MaxPacketPerSec: NumberType().min(0).isRequired(RequiredString),
      KickAvgPacketPerSec: NumberType().min(0).isRequired(RequiredString),
    }),
    FloodThreshholds: ObjectType().shape({
      MaxPacketSize: NumberType().min(0).isRequired(RequiredString),
      MaxPacketPerSec: NumberType().min(0).isRequired(RequiredString),
      KickAvgPacketPerSec: NumberType().min(0).isRequired(RequiredString),
    }),
    MinimumPing: NumberType().min(0).isRequired(RequiredString),
    ErrorMarginFactor: NumberType().min(0).isRequired(RequiredString),
    NaturalLowerMargin: NumberType().min(0).isRequired(RequiredString),
    NaturalUpperMargin: NumberType().min(0).isRequired(RequiredString),
    AllowedSpikePackets: NumberType().min(0).isRequired(RequiredString),
    BaseDesyncForegiveness: NumberType().min(0).isRequired(RequiredString),
    DesyncForgivenessFactor: NumberType().min(0).isRequired(RequiredString),
    DesyncForgivenessInterval: NumberType().min(0).isRequired(RequiredString),
    }),
}),
//non-nested-properties below...
```

In the form I'm following the tutorials and I have a simple input
```typescript
<Form.Group controlId={name}>
    <Form.ControlLabel>
      <div id={name}>{label}</div>
    </Form.ControlLabel>
    <Form.Control
      className="editor-form-field"
      name={name}
      accepter={accepter}
      errorMessage={keyError === name && errorMessage}
      errorPlacement="topStart"
      {...rest}
    />
</Form.Group>
```
When I pass the name, only values that are not nested (or that are not objects) get the DEFAULT VALUE, and when I change them, the values change correctly.

But for objects nothing works, neither the default value is get nor any value is set in onChange, so how can I make them get the default value? and how can I make them correctly define the value in onChange?

ps: i am getting the default values from a json file
non-nested values are ok, but values nested inside an object, nothing works

What value should I pass to "name" so that it gets the object's values and also sets the object's values?
Why is there no tutorial on how to work with objects?
There is tutorial with array, but not with object
How can I set the forms to start with a default value? from an object?
How can I make the forms set the value in the specific property, inside the object?
How to work with objects, within objects, within objects (as in the schema above)?

**Comment by WeylonSantana** [comment]
For those wondering, I still CANNOT find any answers on how to get the initial values of an object and also how to set the value

On the documentation site, in "Form validation Page" I created this below, so if anyone can help with this code below to make the "street and number" inputs get the values from inside the object and set the values inside the object, it would help me a lot, I could get by with a simple answer

```javascript
import { useState } from 'react';
import { Form, Button, Schema, Panel, InputNumber, FlexboxGrid } from 'rsuite';
import JSONTree from 'react-json-tree';
console.clear()

const JSONView = ({ formValue }) => (
  <div style={{ marginBottom: 10 }}>
    <Panel className="json-tree-wrapper" header={<p>formValue</p>}>
      <JSONTree data={formValue} />
    </Panel>
  </div>
);

const model = Schema.Model({
	name: Schema.Types.StringType().isRequired('This field is required.'),
  email: Schema.Types.StringType().isEmail('Please enter a valid email address.'),
  adress: Schema.Types.ObjectType().shape({
		street: Schema.Types.StringType().isRequired('This field is required.'),
  	number: Schema.Types.NumberType().isRequired('This field is required.'),
  })
})

function App() {
  const initialValue = {
    name: 'harry',
    email: 'some@email.com',
    adress: {
      street: 'rt - oh my god',
      number: 45,
    }
  }
  const [formValue, setFormValue] = React.useState(initialValue)
  
  return (
    <Form model={model}
      onChange={(value) => {
        console.log(value)
        setFormValue(value)
      }} 
      formDefaultValue={initialValue}>
      <Form.Group controlId="name">
      	<Form.ControlLabel>Username</Form.ControlLabel>
      	<Form.Control name="name" />
    	</Form.Group>
      
      <Form.Group controlId="email">
      	<Form.ControlLabel>Email</Form.ControlLabel>
      	<Form.Control name="email" />
    	</Form.Group>
      
      <Form.Group controlId="adress">
      	<Form.ControlLabel>Adress</Form.ControlLabel>
      	<Form.Control name="street" />
      	<Form.ControlLabel>Number</Form.ControlLabel>
      	<Form.Control name="number" accepter={InputNumber} />
    	</Form.Group>
      
      <FlexboxGrid.Item colspan={12}>
        <JSONView formValue={formValue} />
      </FlexboxGrid.Item>
      <Button appearance="primary" type="submit">
        Submit
      </Button>
    </Form>
  );
}

ReactDOM.render(<App />, document.getElementById('root'));
```

**Reply by SevenOutman** [reply]
`<Form.Control name="adress.street" />` is supposed to work. But if it's not working, there is a bug then.

**Reply by WeylonSantana** [reply]
> `<Form.Control name="adress.street" />` is supposed to work. But if it's not working, there is a bug then.

It doesn't work, you can copy and paste the code from the first comment above and test it on the "form validation" page

**Reply by SevenOutman** [reply]
@simonguo Does Form.Control support nested paths as `name`?

**Reply by simonguo** [reply]
Yes, it should be supported.

**Reply by WeylonSantana** [reply]
@simonguo you can show a example? In some code sandbox? Getting and setting a value? If you can, please

**Reply by simonguo** [reply]
I mean Form.Control currently doesn't take this usage scenario into consideration, it should support names for nested paths. We'll refine it in later iterations.

To solve your problem at present, another way can also be achieved by customizing form controls.

```tsx
const AddressField = ({ value, defaultValue, onChange, ...rest }) => {
  const [address, setAddress] = useState(value || defaultValue);
  const handleChange = (key, value) => {
    const nextValue = { ...address, [key]: value };
    setAddress(nextValue);
    onChange?.(nextValue);
  };
  return (
    <>
      <Form.ControlLabel> Street </Form.ControlLabel>
      <Input
        value={address?.street}
        onChange={(value) => handleChange("street", value)}
      />
      <Form.ControlLabel>Number</Form.ControlLabel>
      <InputNumber
        value={address?.number}
        onChange={(value) => handleChange("number", value)}
      />
    </>
  );
};

...

<Form.Control name="address" accepter={AddressField} />

```

> Complete example:  https://codesandbox.io/s/lingering-currying-z6grxg

**Reply by WeylonSantana** [reply]
> I mean Form.Control currently doesn't take this usage scenario into consideration, it should support names for nested paths. We'll refine it in later iterations.
> 
> To solve your problem at present, another way can also be achieved by customizing form controls.
> 
> ```tsx
> const AddressField = ({ value, defaultValue, onChange, ...rest }) => {
>   const [address, setAddress] = useState(value || defaultValue);
>   const handleChange = (key, value) => {
>     const nextValue = { ...address, [key]: value };
>     setAddress(nextValue);
>     onChange?.(nextValue);
>   };
>   return (
>     <>
>       <Form.ControlLabel> Street </Form.ControlLabel>
>       <Input
>         value={address?.street}
>         onChange={(value) => handleChange("street", value)}
>       />
>       <Form.ControlLabel>Number</Form.ControlLabel>
>       <InputNumber
>         value={address?.number}
>         onChange={(value) => handleChange("number", value)}
>       />
>     </>
>   );
> };
> 
> ...
> 
> <Form.Control name="address" accepter={AddressField} />
> ```
> 
> > Complete example:  https://codesandbox.io/s/lingering-currying-z6grxg

Got it, I'll do as you said above, so this discussion is open, mark it as an answer?

**Reply by SevenOutman** [reply]
> Got it, I'll do as you said above, so this discussion is open, mark it as an answer?

If this solution works for you, you can mark this comment by yourself as answer, and edit it a bit to let future readers refer to the solution https://github.com/orgs/rsuite/discussions/3100#discussioncomment-5226896.
